docs/data/toolpad/concepts/data-providers.md

```diff
@@ -0,0 +1,117 @@
+
+{{"component": "modules/components/DocsImage.tsx", "src": "/static/toolpad/docs/concepts/data-providers/rows-source.png", "alt": "Select data provider row source", "caption": "Select data provider row source", "zoom": false, "width": 297}}
+
+3. Click the data provider selector and choose **Create new data provider**.
+
+{{"component": "modules/components/DocsImage.tsx", "src": "/static/toolpad/docs/concepts/data-providers/create-data-provider.png", "alt": "Create data provider", "caption": "Create data provider", "zoom": false, "width": 294}}
+
+4. Name the new data provider and click **Create**
+
+{{"component": "modules/components/DocsImage.tsx", "src": "/static/toolpad/docs/concepts/data-providers/create-data-provider-dialog.png", "alt": "Create data provider dialog", "caption": "Create data provider dialog", "zoom": false, "width": 490}}
+
+5. Use the code button to open your code editor with the data provider backend.
+
+{{"component": "modules/components/DocsImage.tsx", "src": "/static/toolpad/docs/concepts/data-providers/open-editor.png", "alt": "Open data provider editor", "caption": "Open data provider editor", "zoom": false, "width": 272}}
+
+6. A data provider that iterates over a static list could look as follows:
+
+   ```tsx
+   import { createDataProvider } from '@mui/toolpad-core/server';
+   import DATA from './movies.json';
+
+   export default createDataProvider({
+     async getRecords({ paginationModel: { start = 0, pageSize } }) {
+       const records = DATA.slice(start, start + pageSize);
+       return { records, totalCount: DATA.length };
+     },
+   });
+   ```
+
+## Pagination
+
+The data provider supports two styles of pagination. Index based, and cursor based pagination.
+
+### Index based
+
+This is the strategy your data is paginated by when it returns data based on a page number and page size. The `getRecords` method will receive `page` and `pageSize` values in it `paginationModel` parameter and returns a set of records representing the page. Index based pagination is the default but you can explicitly enable this by setting `paginationMode` to `'index'`.
+
+```tsx
+export default createDataProvider({
+  paginationMode: 'index',
+  async getRecords({ paginationModel: { start = 0, pageSize } }) {
+    const { page, totalCount } = await db.getRecords(start, pageSize);
+    return {
+      records: page,
+      totalCount,
+    };
+  },
+});
+```
+
+### Cursor based
+
+This is the strategy your data is paginated by when it returns data based on a cursor and a page size. The `getRecords` method will receive `cursor` and `pageSize` values in it `paginationModel` parameter and returns a set of records representing the page. You indicate the cursor of the next page with a `cursor` property in the result. Pass `null` to signal the end of the collection. The You can enable Cursor based pagination by setting `paginationMode` to `'cursor'`.
+
+```tsx
+export default createDataProvider({
```

Comment:
Maybe i'm not super familiar with cursor pagination, but it isn't completely obvious to me right away how to use this cursor pagination.
is the cursor a number? Or can I use the id of a record?
The prisma documentation on this kind of this is very clear, maybe we could use it as inspiration https://www.prisma.io/docs/concepts/components/prisma-client/pagination.

Reply:
Oh, also the way Prisma does it kind of just works and we don't have to explicitly choose a pagination model. Not sure if we could do something similar?

Reply:
Nevermind I understand this is quite different from Prisma now as it integrates very directly with the Data Grid, these were just some quick ideas while going over the docs and trying to use the feature in a basic way. Feel free to ignore if they don't make sense or don't provide any insights :D

Reply:
I've added another paragraph explaining the value of `cursor` in this pagination model.

> Oh, also the way Prisma does it kind of just works and we don't have to explicitly choose a pagination model. Not sure if we could do something similar?

In prisma you don't need to explicitly specify a pagination model because it doesn't need to signal to consumers how it is to be called. The Toolpad data provider needs to signal to Toolpad how it's to be fetched. Supplying this option configures the interfaces in `getRecords` for the right pagination model.